pyro/infer/svi.py

```diff
@@ -37,11 +38,17 @@ def __init__(self,
         self.optim = optim
 
         if isinstance(loss, ELBO):
-            self.ELBO = loss
-            self.loss = self.ELBO.loss
-            self.loss_and_grads = self.ELBO.loss_and_grads
+            self.loss = loss.loss
+            self.loss_and_grads = loss.loss_and_grads
         else:
-            raise TypeError("Unsupported loss type. Expected an ELBO instance, got a {}".format(type(loss)))
+            if loss_and_grads is None:
+                def _loss_and_grads(*args, **kwargs):
+                    loss_val = loss(*args, **kwargs)
+                    loss_val.backward()
+                    return loss_val
```

Comment:
In `ELBO` implementations, `.loss()` returns a float rather than a tensor. This makes it difficult to support automatic definition of `.loss_and_grads()`.

What do you think of defining a `Loss` base class from which `ELBO` inherits, and requiring that base class to implement either `.differentiable_loss` or both of `.loss()` and `.loss_and_grads()`? This would allow us to refactor `TraceEnum_ELBO` to expose our Dice loss:
```py
class Loss(object):
    def differentiable_loss(self, *args, **kwargs):
        raise NotImplementedError
    def loss(self, *args, **kwargs):
        return self.diffentiable_loss(*args, **kwargs).item()
    def loss_and_grads(self, *args, **kwargs):
        loss = self.differentiable_loss
        loss.backward()  # TODO support create_graph=True
        return loss.item()

class ELBO(Loss):
    # ...unchanged...

class TraceEnum_ELBO(ELBO):
    def differentiable_loss(self, *args, **kwargs):
        # ...simplified...
```
This seems like it also paves the way to supporting higher-order optimizers. Thoughts?

Reply:
> In ELBO implementations, .loss() returns a float rather than a tensor.

Why do we need to do this instead of just using `with torch.no_grad()`?

Reply:
WDYT of the `Loss` base class that would help simplify `TraceEnum_ELBO`?

There are a couple reasons we're currently returning float-valued loss:
1. floats are easier to deal with when e.g. printing, and that's all that can be done with them by the time they're returned from a `.loss()` function (`.loss()` need not guarantee differentiability, and `.loss_and_grads()` need not create an autograd graph)
2. python floats have higher precision (64 bits) than typical tensors (32bits), so we can simplify some of the aggregation in `Trace_ELBO`

Reply:
3. users would often forget to call `.detach()`. They then run out of memory in the training loop via `losses.append(svi.step())`, as autograd graphs are never released. If we support custom loss functions w/o a `Loss` base class, we should at least make `SVI.step()` safer by
    ```py
    return loss.item() if torch.is_tensor(loss) else loss
    ```

Reply:
> WDYT of the Loss base class that would help simplify TraceEnum_ELBO?

Just wrapped the return values of `SVI.step` and `SVI.evaluate_loss` with `torch_item`, should that accomplish the same thing?